Add ENS support for addressing ERC721 NFTs (#2381)
ethereum/EIPs: EIPS/eip-2381.md

---
eip: 2381
title: ENS support for EIP-721 TokenID
author: Oisin Kyne (@OisinKyne)
discussions-to: https://github.com/ethereum/EIPs/issues/2381
status: Draft
type: Standards Track
category: ERC
created: 2019-11-15
requires: 137, 165, 721
---

## Simple Summary

> This EIP makes [EIP-721](https://eips.ethereum.org/EIPS/eip-721) Non-Fungible Tokens addressable by the Ethereum Name Service ([EIP-137](https://eips.ethereum.org/EIPS/eip-137)) by adding a new resolver profile, `uint256 tokenID()`. This allows the Ethereum Name Service to address a single non-fungible token within an EIP-721 token contract of many tokens.

Reviewer: This is still longer than we would like.  Think email subject line, or Git commit message header, or the last column in a table.

Author: Cut it down significantly and moved things to motivation

## Abstract

It makes sense to resolve an ENS name to just a contract address for fungible tokens such as [EIP-20](https://eips.ethereum.org/EIPS/eip-20), as each token in the contract is indistinguishable from another. However, for non-fungible tokens; pointing to the contract address alone is not enough, as tokens within the contract are meant to be unique and distinguishable. Non-fungibles might have different valuations, might grant the holder different rights or rewards etc.

A single deployed EIP-721 contract can contain multiple non-fungible tokens. Tokens are addressed by an unsigned 256bit integer, `tokenID`. This EIP describes an extra getter and setter, `tokenID()` and `setTokenID()`, that are included in an ENS resolver contract, to allow for ENS names to resolve to a specific non fungible token within a given EIP-721 contract.

> This allows ENS domains like [`bugcat.cryptokitties.eth`](https://www.cryptokitties.co/kitty/101) and [`dragon.cryptokitties.eth`](https://www.cryptokitties.co/kitty/896775) to resolve to both the CryptoKitties contract address and the tokenID of the non-fungible kitty within it.

Reviewer: This content all seems like motivation.  Abstract should be a human readable version of the core of the technical specification.  Something along the lines of:
> Creates a standard ENS resolver interface that contains the methods `tokenID(node)` and `setTokenID(node, token)`.  These methods allow an ENS user to set/query...

Author: Moved the non technical aspects to motivation

## Motivation

Non-fungible Tokens are meant to be globally unique, it makes sense to be able to name them. Giving NFTs names makes them more real to the non Ethereum enthusiast that doesn't understand what a contract address or tokenID is. A domain name in the existing web 2.0 world is already widely understood to be a finite resource, and as such, would help convey the scarcity that is a non-Fungible token to new users.

This change might also to an extent, decentralise access to non-fungibles on Ethereum. Currently, the mapping for a human readable name to an NFT is kept off chain, typically on the platform of the NFT issuer. However, if the community moved towards naming NFTs on chain, using ENS, this would allow any client that supports ENS resolution to resolve an ENS name to an NFT, without querying the issuer's off-chain metadata.

## Specification

The ENS Resolver Profile specified by this EIP adds two functions to a deployed resolver contract:

```

>     function tokenID(bytes32 node) public view returns(uint256);
>     function setTokenID(bytes32 node, uint256 token);

Reviewer: ```suggestion
function tokenID(bytes32 node) public view returns(uint256);
function setTokenID(bytes32 node, uint256 token);
```

>     function setTokenID(bytes32 node, uint256 token);

Reviewer: The specification should probably specify what value `tokenId` contains.  At the moment, if I use this resolver and call `setTokenID(..., 101)` and then someone resolves via `tokenID(...)`, they will not have enough information to get to `https://www.cryptokitties.co/kitty/101` as the motivation suggests.  This implies to me that there is something left unspecified here.

Author: Added an explainer to line 21 that the URL is to be retrieved from from EIP721's `tokenURI` functionality, and this EIP is just to give people the tokenID they need to request the right url. https://github.com/ethereum/EIPs/blame/master/EIPS/eip-721.md#L193 hope that makes it clearer.

```

> You can evaluate whether a given ENS resolver supports this EIP by using the EIP-165 standard `supportsInterface(bytes4)`.

Reviewer: ```suggestion
You can evaluate whether a given ENS resolver supports this EIP by using the [EIP-165](./eip-165.md) standard `supportsInterface(bytes4)`.
```

The interface identifier for this interface is `0x4b23de55`.

## Rationale

The design of this smart contract is a modified replica of the [existing contenthash resolver profile](https://github.com/ensdomains/resolvers/blob/master/contracts/profiles/ContentHashResolver.sol).

The decision to use Contract Address and tokenID as a pair to uniquely identify EIP-721 NFTs is lifted directly from the [EIP-721 Rationale](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md#rationale) quoted here:

>     Every NFT is identified by a unique uint256 ID inside the EIP-721 smart contract. This identifying number SHALL NOT change for the life of the contract. The pair (contract address, uint256 tokenId) will then be a globally unique and fully-qualified identifier for a specific asset on an Ethereum chain.

Reviewer: ```suggestion
> Every NFT is identified by a unique uint256 ID inside the EIP-721 smart contract. This identifying number **SHALL NOT** change for the life of the contract. The pair (contract address, uint256 tokenId) will then be a globally unique and fully-qualified identifier for a specific asset on an Ethereum chain.
```
I'm not confident that an indent will result in a quote in all rendering engines, but I'm pretty sure `>` will.

---
eip: 2381
title: ENS support for ERC 721 TokenID
author: Oisin Kyne (@OisinKyne) <oisin@kyne.eu>
discussions-to: https://github.com/ethereum/EIPs/issues/2381

> status: Last Call

Reviewer 2: This isn't even merged as a draft. It can't start out as last call.

Author: Sorry, I didn't realise they merge as draft, I figured the PR being merged was the last step. Will revert.

> type: Meta

Reviewer: ```suggestion
type: Interface
```
Meta is for specifications about EIPs and the process itself.

Reviewer 2: I think this is an ERC.

Author: I think @axic has it right. Changed to `ERC`.

created: 2019-11-15

> requires (*optional): EIP137, EIP165, EIP721

Reviewer 2: These are badly formatted, should be: `requires: 137, 165, 721`.

---

<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->

## Simple Summary

<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->

> _TL;DR_: This EIP makes ERC721 Non-Fungible Tokens addressable by the Ethereum Name Service by adding a new resolver profile, `uint256 tokenID()`.

Reviewer: ```suggestion
This EIP makes [ERC-721](./eip-721.md) Non-Fungible Tokens addressable by the Ethereum Name Service by adding a new resolver profile, `uint256 tokenID()`.
```
The first reference to any EIP in an EIP should be a link to that EIP.  Also, the correct formatting for EIP/ERCs is `EIP-XXX` (note the hyphen).  That needs to be fixed throughout this document.

Author: Fixed throughout. Used the FQDN (e.g. https://eips.ethereum.org/EIPS/eip-20) rather than a relative link within the repo if that's acceptable. Can change.

Reviewer: They do need to be relative links.  This ensures that they work when viewing in GitHub, on EIPs repository, in web archive, on a mirror, etc.

_Slightly longer TL;DR_: It makes sense to resolve an ENS name to a contract address for fungible tokens such as ERC20, as each token in the contract is indistinguishable from another. However, for non-fungible tokens, pointing to the contract address alone is not enough, as tokens within the contract are meant to be unique and distinguishable. Non-fungibles might have different valuations, might grant the holder different rights or rewards etc.

> This EIP proposes a new ENS resolver for ERC721's tokenID field. This allows the Ethereum Name Service to address a single non-fungible token within an ERC721 token contract of many tokens.

Reviewer: Simple Summary should just be the TL;DR (but without that prefix).  It is meant to be a one-line description of what this standard is about.  The abstract is where you would put a more verbose description, usually a human-readable and terse version of the specification.

Also, EIPs should be worded as though read by a future reader.  This means that EIPs do not propose, they assert.  "This EIP describes a new ENS resolver for ERC721's tokenID field".

Author: TL;DRs removed and EIP reworded to sound post-fact rather than like a proposal.

## Further Questions

### Can I use this EIP for my other token standard that uses a uint256 to address tokens?

Yes, I have no problem using this resolver for any token contract that addresses individual tokens with a `uint256` number. However, I would ask that you formalise the adoption of this resolver for your token contract in an EIP and reference this EIP 2381 in it.

### How do you know when to check if an ENS name supports this EIP?

To make sure everything is as expected, it would be good to check that:

- The resolver contract for a given ENS name supports this EIP's (2381) interface (`0x4b23de55`)
- The contract address pointed at by the resolver of an ENS name implements ERC721 (Interface ID of: `0x80ac58cd`)

> - One specific edge case to be aware of. In the ERC721 spec, a `tokenID` of 0 is a valid identifier of a token. If an ENS name uses this new resolver contract, but doesn't set a `tokenID` value, this field will default to 0, which implicitly is a valid `tokenID`. If an ERC721 contract owner wants to address their contract with an ENS name, but does not want to address any specific token within it, they should use a resolver contract without this EIP included, to avoid the case where an ENS resolver infers that this name is to point to `tokenID` 0 within the contract, which may or may not exist. The alternative for this edge case would be to make this EIP require that `tokenID` be non-zero to be valid, but this would make specific tokens with `tokenID` of 0 not addressable by this EIP, which would not be ideal, considering they are compliant tokens in the 721 spec.

Author: Cut a question out and rewrote the other and moved it to part of the implementation section to explain how to interact with the EIP implementation as spec'd

> ## Copyright

Author: I wrote a section for Security Considerations, it could probably be better though. Advice welcome. https://github.com/OisinKyne/EIPs/blob/master/EIPS/eip-2381.md#security-considerations

> ## Copyright

Reviewer: A security considerations section could just read "there are no known security issues associated with this EIP", it doesn't have to be complicated.  :)  In this case, I think you have some good info in here though that is worth mentioning.
